### PR TITLE
New module: Waiting for VMware tools to become available

### PR DESCRIPTION
**Pull request by pdellaert**
##### SUMMARY
Adding a new module which supports waiting for VMware tools to become available for a VMware VM (VMware tools in a running state)

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
vmware_guest_tools

##### ANSIBLE VERSION
```
ansible 2.4.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/pdellaer/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/pdellaer/Virtualenvs/ansible/lib/python2.7/site-packages/ansible
  executable location = /home/pdellaer/Virtualenvs/ansible/bin/ansible
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```

##### ADDITIONAL INFORMATION
govcsim does not support mimicking the behaviour of VMware tools, so integration tests are not created at this moment.

### CODE REVIEW

**Comment by nerzhul** [comment]
Sounds interesting idea for a preseed deployment or template deployment. If i'm correct you should use get_obj from vmware utils (ping @Akasurde)

**Comment by pdellaert** [comment]
@nerzhul Use case is a situation where something is deployed either through ovftool, or cloned from a template (for instance if wait_for_ip is not an option for some reason), and you need to wait on VMware tools to come up before you can push commands to the VM.

*update*
For finding the VM, i looked at existing modules, specifically to `vmware_guest_facts`, as i wanted to stay similar to existing `vmware_guest_*` specs. Changing that to functionality from the vmware utils, is a possibility. It does seem the available methods for finding a VM by name in the utils, does not take into account the 'first' or 'last' functionality from the `vmware_guest_*` modules

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+class PyVmomiHelper(object):
+    def __init__(self, module):
+        if not HAS_PYVMOMI:
+            module.fail_json(msg='pyvmomi module required')
+
+        self.module = module
+        self.params = module.params
+        self.content = connect_to_api(self.module)
+
+    def getvm(self, name=None, uuid=None, folder=None):
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools.py]: Can you use refactored [getvm](https://github.com/ansible/ansible/pull/27188/files#diff-d5ca044435f1fc628995b038d2e238f7R173) ?

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        if not tools_running:
+            return {'failed': True, 'msg': 'VMware tools either not present or not running after {0} seconds'.format((poll * sleep))}
+
+        changed = False
+        if poll_num > 0:
+            changed = True
+        return {'changed': changed, 'failed': False, 'instance': vm_facts}
+
+
+def get_obj(content, vimtype, name):
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools.py]: get_obj is unused, can be removed.

File: test/integration/targets/vmware_guest_tools/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+#- debug: msg="{{ guest_facts_0001 }}"
+#
+#- assert:
+#    that:
+#      - "guest_facts_0001['instance']['hw_name'] == vm1 | basename"
+#      - "guest_facts_0001['instance']['hw_product_uuid'] is defined"
+#      - "guest_facts_0001['instance']['guest_tools_status'] == 'guestToolsRunning'"
+#
+#- set_fact: vm1_uuid="{{ guest_facts_0001['instance']['hw_product_uuid'] }}"
+#
```
Comment (Akasurde) [inline test/integration/targets/vmware_guest_tools/tasks/main.yml]: LGTM

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools.py (hunk 1 above)
Comment (pdellaert) [reply lib/ansible/modules/cloud/vmware/vmware_guest_tools.py]: Thanks, done

File: test/integration/targets/vmware_guest_tools/tasks/main.yml (hunk 3 above)
Comment (pdellaert) [reply test/integration/targets/vmware_guest_tools/tasks/main.yml]: thanks

**Comment by pdellaert** [comment]
@Akasurde all suggestions were implemented, thanks for the review!

**Review by Akasurde — APPROVED** [review]
shipit

**Comment by pdellaert** [comment]
@Akasurde @jctanner 
For changelog, i'd like to add this module
For BOTMETA, i can either add myself to $team_vmware, or just add myself as the maintainer, up to you?

**Comment by dagwieers** [comment]
So the module freeze deadline for v2.4 is getting closer (2017-08-29). Now is the time to finish up, get it reviewed and merged with no delay.

**Comment by pdellaert** [comment]
@Akasurde @bcoca @jctanner Can you review/merge?

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: vmware_guest_tools_wait
+short_description: Wait for VMware tools to become available and return facts
+description:
+    - Wait for VMware tools to become available on the VM and return facts
+version_added: 2.4
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: This should be a string.

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: vmware_guest_tools_wait
+short_description: Wait for VMware tools to become available and return facts
+description:
+    - Wait for VMware tools to become available on the VM and return facts
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: Requires a trailing dot.

**Review by Akasurde — APPROVED** [review]
shipit

**Comment by dav1x** [comment]
shipit

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# This file is part of Ansible
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: We have a new one-line GPL message that replace the below header.

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    - Philippe Dellaert (@pdellaert) <philippe@dellaert.org>
+notes:
+    - Tested on vSphere 6.5
+requirements:
+    - python >= 2.6
+    - PyVmomi
+options:
+   name:
+        description:
+            - Name of the VM for which to wait until the tools become available
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: Descriptions (except short_description) require a trailing dot.

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+            - This is required if uuid is not supplied.
+   name_match:
+        description:
+            - If multiple VMs matching the name, use the first or last found
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: Descriptions (except short_description) require a trailing dot.

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        default: 'first'
+        choices: ['first', 'last']
+   folder:
+        description:
+            - Destination folder, absolute or relative path to find an existing guest.
+            - This is required if name is supplied.
+            - The folder should include the datacenter. ESX's datacenter is ha-datacenter
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: Descriptions (except short_description) require a trailing dot.

Comment (dagwieers) [reply lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: Maybe do `C(ha-datacenter)` so it stands out better.

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+            - This is required if uuid is not supplied.
+   name_match:
+        description:
+            - If multiple VMs matching the name, use the first or last found
+        default: 'first'
+        choices: ['first', 'last']
+   folder:
+        description:
+            - Destination folder, absolute or relative path to find an existing guest.
+            - This is required if name is supplied.
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: Please do `C(name)` to make parameters stand out.

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+            - '   folder: /datacenter1/vm/folder1'
+            - '   folder: datacenter1/vm/folder1'
+            - '   folder: /folder1/datacenter1/vm'
+            - '   folder: folder1/datacenter1/vm'
+            - '   folder: /folder1/datacenter1/vm/folder2'
+        default: /vm
+   uuid:
+        description:
+            - UUID of the VM  for which to wait until the tools become available, if known. This is VMware's unique identifier.
+            - This is required if name is not supplied.
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: Same here, do `C(name)`

**Review by dagwieers — CHANGES_REQUESTED** [review]
Apart from the (pedantic) cosmetics / documentation bits, LGTM.

**Comment by pdellaert** [comment]
@dagwieers Thanks for the review, implemented your comments.

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+    module.params['folder'] = module.params['folder'].rstrip('/')
+
+    pyv = PyVmomiHelper(module)
+    # Check if the VM exists before continuing
+    vm = pyv.getvm(name=module.params['name'],
+                   folder=module.params['folder'],
+                   uuid=module.params['uuid'])
+
+    # VM exists
+    if vm:
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: I would also do this:
```python
if not vm:
    # fail

# continue with the module
```
This makes it easier to read, as you don't have to find what `else` is part of what condition.

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+                module.exit_json(**result)
+        except Exception as exc:
+            module.fail_json(msg="Waiting for tools failed with exception: %s" % to_native(exc))
+    else:
+        msg = "Unable to wait for tools for non-existing VM "
+        if module.params['name']:
+            msg += "%(name)s" % module.params
+        elif module.params['uuid']:
+            msg += "%(uuid)s" % module.params
+        module.fail_json(msg=msg)
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: You could actually make this fancier:
```python
    vmid = module.params.get('name') or module.params.get('uuid')
    module.fail_json(msg="Unable to wait for tools for non-existing VM '%s'" % vmid)
```

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+    # Check if the VM exists before continuing
+    vm = pyv.getvm(name=module.params['name'],
+                   folder=module.params['folder'],
+                   uuid=module.params['uuid'])
+
+    # VM exists
+    if vm:
+        try:
+            result = pyv.wait_for_tools(vm)
+            if result['failed']:
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: Don't add stuff in a try-except block that should not fail.
If possible also catch specific exceptions with specific error messages.
(It is possible not all exceptions provide a workable string the same way, unfortunately)

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: vmware_guest_tools_wait
+short_description: Wait for VMware tools to become available and return facts
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: I would make the short_description shorter by removing the "and return facts". This description is used in the index, so we don't need those details.

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+import time
+
+# import module snippets
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+from ansible.module_utils.vmware import connect_to_api, gather_vm_facts, vmware_argument_spec, find_vm_by_id
+
+try:
+    import json
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: I learned yesterday that nowadays we can simply do `import json` as we no longer support Python 2.4

File: lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+instance:
+    description: metadata about the virtual machine
+    returned: always
+    type: dict
+    sample: None
+"""
+
+import time
+
+# import module snippets
```
Comment (dagwieers) [inline lib/ansible/modules/cloud/vmware/vmware_guest_tools_wait.py]: This comment does not offer anything useful, kill it !
(It's an old legacy thing)

**Review by dagwieers — CHANGES_REQUESTED** [review]
Almost there !

**Comment by pdellaert** [comment]
Closing and reopening as my latest commit does not show up in PR...

**Comment by pdellaert** [comment]
Thanks @dagwieers for the detailed check! On the json import: i just removed it, as it turned out not to be needed (which was your intention, of course ;)), should have caught that initially.

**Review by dagwieers — APPROVED** [review]

**Pull request merged**